SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp

```diff
@@ -0,0 +1,254 @@
+  const std::string m_is_source_or_dst_simd;
+};
+
+// The main test routine.
+// Using functor class to be able to iterate over the pre-defined data types.
+template <typename DataT, typename SizeT, typename TestCaseT> class run_test {
+  static constexpr int NumElems = SizeT::value;
+
+public:
+  bool operator()(sycl::queue &queue, const std::string &data_type) {
+    bool passed = true;
+    std::vector<DataT> ref_data = generate_ref_data<DataT, NumElems>();
+
+    if constexpr (TestCaseT::is_increment()) {
+      mutate(ref_data, mutator::For_addition<DataT>(1));
+    } else if constexpr (!TestCaseT::is_increment()) {
+      mutate(ref_data, mutator::For_subtraction<DataT>(1));
+    }
+
+    // If current number of elements is equal to one, then run test with each
+    // one value from reference data.
+    // If current number of elements is greater than one, then run tests with
+    // whole reference data.
+    if constexpr (NumElems == 1) {
+      for (size_t i = 0; i < ref_data.size(); ++i) {
+        passed &= run_verification(queue, {ref_data[i]}, data_type);
+      }
+    } else {
+      passed &= run_verification(queue, ref_data, data_type);
+    }
+    return passed;
+  }
+
+private:
+  bool run_verification(sycl::queue &queue, const std::vector<DataT> &ref_data,
+                        const std::string &data_type) {
+    assert(ref_data.size() == NumElems &&
+           "Reference data size is not equal to the simd vector length.");
+
+    bool passed = true;
+
+    shared_allocator<DataT> allocator(queue);
+    shared_vector<DataT> source_simd_out(NumElems, allocator);
+    shared_vector<DataT> result_simd_out(NumElems, allocator);
+    shared_vector<DataT> shared_ref_data(ref_data.begin(), ref_data.end(),
+                                         allocator);
+
+    queue.submit([&](sycl::handler &cgh) {
+      const DataT *const ref = shared_ref_data.data();
+      DataT *source_simd_data_storage = source_simd_out.data();
+      DataT *result_simd_data_storage = result_simd_out.data();
+
+      cgh.single_task<Kernel<DataT, NumElems, TestCaseT>>(
+          [=]() SYCL_ESIMD_KERNEL {
+            TestCaseT::template call_simd_ctor<DataT, NumElems>(
+                ref, source_simd_data_storage, result_simd_data_storage);
+          });
+    });
+    queue.wait_and_throw();
+
+    for (size_t i = 0; i < NumElems; ++i) {
+
+      DataT expected_value = shared_ref_data[i];
+      DataT source_simd_expected_value = expected_value;
+
+      if constexpr (TestCaseT::is_increment()) {
+        ++source_simd_expected_value;
+      } else {
+        --source_simd_expected_value;
+      }
+
+      passed &= verify_result(i, source_simd_expected_value, source_simd_out[i],
+                              "result of calling", data_type);
+
+      expected_value = TestCaseT::apply_operator(expected_value);
+      passed &= verify_result(i, expected_value, result_simd_out[i],
+                              "return value for", data_type);
```

Comment:
Would you mind to provide an example of failure output for both `verify_result` calls?
Anyway, I would suggest to use something regarding "specific check" instead of `source_or_destination_simd` and `context` variable names.

Reply:
We will have the same output like this: Failed for simd<sycl::half, 8>, retrieved: -nan, expected: -nan(ind), at index: 1 for simd as a return value for pre increment operator.

> Anyway, I would suggest to use something regarding "specific check" instead of source_or_destination_simd and context variable names.

I will think about the new names

Reply:
Agree, I would suggest something regarding `check_description` though. The `simd_type` is more like the `simd<DataT, NumElems>` itself.

Reply:
@yuriykoch which name you will suggest instead of current name?

Reply:
What about the following instead of `check_description` or `specific_check`:
```
      passed &= verify_result(i, expected_source_value, source_simd_out[i],
                              "unexpected argument modification", data_type);
      passed &= verify_result(i, expected_return_value, result_simd_out[i],
                              "unexpected return value", data_type);

....

    log_msg += ", at index: " + std::to_string(m_index);
    log_msg += "for " + TestCaseT::get_description() + " operator: ";
    log_msg += m_error_details;
```

Reply:
new error messages will be looks like:
```
Failed for simd<float, 32>, retrieved: 4.100000, expected: 4.100000, at index: 31 for post increment operator: unexpected argument modification
Failed for simd<float, 32>, retrieved: 3.100000, expected: 3.100000, at index: 31 for post increment operator: unexpected return value
```
I think that these is ok messages